### PR TITLE
5.x

### PR DESCRIPTION
**Pull request by fetzig**
tracking pull request for unified SDK

TODOS:
* do all listed code changes (see below)
* **migrate 4.x improvements that happend since 5.x start to 5.x (tvOS fix #373, swift package manager, a.s.o.)**
* check all TODOs in code. fix the easy ones, get rid of deprecated ones, and make sure the rest is easy to understand.
* validate if 5.x conforms with unified API
* fix all tests and add new ones to get code coverage up again
  * analyze and profile (mem alloc/leaks, performance)
* update getsentry/examples branch `5.x`
* check oclint/lint

Code Changes: 

- [x] `setLogLevel` in `SentryClient` shouldn't exist. Instead, there should be an option in `debug`: `boolean` which enables the SDK to log Verbose or default should be Error level. Ideally, the debug option also support `boolean` but also logLevel as string `none | error | verbose | debug` (default should be error, none and verbose are cocoa sdk specific - maybe deprecate at some point).
- [x] `versionString` and `sdkName` shouldn't live in `SentryClient` it should be in a separate file so we can easily bump version in a seperate file. (add it in a new/separate file)
- [x] `Scope` needs `setContext` (see sentry-js)
- [x] `systemInfo` in `SentryContext` is dependend on `SentryCrash` this should become an intergration (We may need eventProcessors)
- [x] `SentryEvent` should not have manual release tracking 
```
if (nil == self.releaseName && nil == self.dist && nil != self.infoDict) {
        self.releaseName = [NSString stringWithFormat:@"%@-%@", self.infoDict[@"CFBundleIdentifier"], self.infoDict[@"CFBundleShortVersionString"]];
        self.dist = self.infoDict[@"CFBundleVersion"];
    }
``` 
this should live in `SentryClient` (prepareEvent) behind an option, default should be enabled. 

things below have been moved to the backlog:

- [ ] ~~captureEnvelope -> We need to implement the concept of envelopes and relay the raw envelope through the transport an send it off. This is used for higher level SDKs like RN, Cordova, Flutter where these SDKs create the event, serialize it as JSON and the "Native" SDK (in that case sentry-cocoa) just relays the Event/envelope to Sentry.~~
```
- (instancetype)initWithJSON:(NSData *)json {
    self = [self initWithLevel:kSentrySeverityInfo];
    if (self) {
        self.json = json;
    }
    return self;
}
```
~~can be removed from `SentryEvent` after implementing this.~~
- [ ] ~~Fix `SentryJavaScriptBridgeHelper` for RN/Cordova~~

### CODE REVIEW

**Comment by tmm1** [comment]
> * fix all tests and add new ones to get code coverage up again
>   * analyze and profile (mem alloc/leaks, performance)

Was memory usage of this branch analyzed?

We're seeing an increase in OOM kills of our app since we upgraded to Sentry 5.0.0-beta.3

File: Sources/Sentry/SentryBreadcrumbStore.m [hunk 1]
```diff
@@ -1,77 +0,0 @@
...
-    if (self) {
-        self.maxBreadcrumbs = 50;
-        self.fileManager = fileManager;
-    }
-    return self;
-}
-
-- (void)addBreadcrumb:(SentryBreadcrumb *)crumb {
-    [SentryLog logWithMessage:[NSString stringWithFormat:@"Add breadcrumb: %@", crumb] andLevel:kSentryLogLevelDebug];
-    [self.fileManager storeBreadcrumb:crumb maxCount:self.maxBreadcrumbs];
```
Comment (tmm1) [inline Sources/Sentry/SentryBreadcrumbStore.m:44]: Previously, only 50 breadcrumbs were stored and they were saved into the filemanager (which presumably stores things on disk).

File: Sources/Sentry/SentryScope.m [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    }
+    return self;
+}
+
+#pragma mark Global properties
+
+- (void)addBreadcrumb:(SentryBreadcrumb *)crumb {
+    [SentryLog logWithMessage:[NSString stringWithFormat:@"Add breadcrumb: %@", crumb] andLevel:kSentryLogLevelDebug];
+    @synchronized (self) {
+        [self.breadcrumbArray addObject:crumb];
```
Comment (tmm1) [inline Sources/Sentry/SentryScope.m:125]: Now, all breadcrumbs are stored indefinitely into RAM causing a memory leak.

Comment (bruno-garcia) [reply Sources/Sentry/SentryScope.m:125]: Yes, this sadly got merged in here: https://github.com/getsentry/sentry-cocoa/pull/348/files#diff-79099d807271a5ec576d48da75162a5fL50

We'll get this fixed, thanks @tmm1 for spotting this.

Comment (bruno-garcia) [reply Sources/Sentry/SentryScope.m:125]: We merged a fix and will be releasing 5.0.0-beta.5 soon:

https://github.com/getsentry/sentry-cocoa/pull/450

**Pull request merged**